river/misc/cms.py

```diff
@@ -0,0 +1,189 @@
+    10000
+
+    We can decrease the error by allocating more memory in the CMS:
+
+    >>> cms_a = misc.CountMin(epsilon=0.001, delta=0.01, seed=0)
+    >>> for v in vals:
+    ...     cms_a.update(v)
+
+    >>> cms_a[7]
+    5
+    >>> cms_a[532]
+    4
+
+    We can also obtain estimates of the dot product between two instances of `CountMin`. This could be useful, for instance,
+    to estimate the cosine distance between the data monitored in two different instances of `CountMin`. Suppose we create another
+    CMS instance (the number of slots and hash tables must match) that monitors another sample of the same data generating process:
+
+    >>> cms_b = misc.CountMin(epsilon=0.001, delta=0.01, seed=7)
+
+    >>> for _ in range(10000):
+    ...     v = rng.randint(-1000, 1000)
+    ...     cms_b.update(v)
+
+    Now, we can define a cosine distance function:
+
+    >>> def cosine_dist(cms_a, cms_b):
+    ...     num = cms_a @ cms_b
+    ...     den = math.sqrt(cms_a @ cms_a) * math.sqrt(cms_b @ cms_b)
+    ...     return num / den
+
+    And use it to calculate the cosine distance between the elements monitored in `cms_a` and `cms_b`
+
+    >>> cosine_dist(cms_a, cms_b)
+    0.17536377918173912
+
+    References
+    ----------
+    [^1]: [Cormode, G., & Muthukrishnan, S. (2005). An improved data stream summary: the count-min sketch and its applications. Journal of Algorithms, 55(1), 58-75.](https://www.cse.unsw.edu.au/~cs9314/07s1/lectures/Lin_CS9314_References/cm-latin.pdf)
+    [^2]: [Count-Min Sketch](https://florian.github.io//count-min-sketch/)
+    [^3]: [Hash functions family generator in Python](https://stackoverflow.com/questions/2255604/hash-functions-family-generator-in-python)
+
+    """
+
+    def __init__(self, epsilon: float = 0.1, delta: float = 0.05, seed: int = None):
+        self.epsilon = epsilon
+        self.delta = delta
+        self.seed = seed
+
+        self._w = int(math.ceil(math.e / self.epsilon))
+        self._d = int(math.ceil(math.log(1 / self.delta)))
+
+        self._rng = random.Random(self.seed)
+        self._masks = [self._rng.getrandbits(64) for _ in range(self._d)]
+        self._cms = np.zeros((self._d, self._w), dtype=np.int32)
+
+    def _hash(self, x):
+        return tuple(zip(*((i, (hash(x) ^ self._masks[i]) % self._w) for i in range(self._d))))
+
+    def __getitem__(self, x) -> int:
+        # Point query
+        return min(self._cms[self._hash(x)])
+
+    def __matmul__(self, other: "CountMin") -> int:
+        # Dot product
+        if self.n_slots != other.n_slots or self.n_tables != other.n_tables:
+            unmatched_dims = (
+                f"({self.n_slots}, {self.n_tables}) vs. ({other.n_slots}, {other.n_tables})"
+            )
+            raise ValueError(f"The number of slots and hash tables do not match: {unmatched_dims}.")
+
+        return min(np.einsum("ij,ij->i", self._cms, other._cms))
+
+    def __len__(self):
+        return self.n_slots * self.n_tables
```

Comment:
Can you add a comment to indicate what this corresponds to?

Reply:
Done!